mcs, tso, client: After NewClientWithAPIContextV2 returns, the keyspace group should be discovered by the passed keyspace name immediately.

## Enhancement Task
### Summary
It's a pure client issue during tso service discovery at the first time. During the incident, we restarted PD pods and TSO pods, but they won't help. The real action which mitigated the issue was we moved group2's primary to the same location of group 0's primary around 13:55, which I'll explain later.
#### What's the problem
After TiDB calls NewClientWithAPIContextV2 and returns, the first GetTS always uses the default keyspace group (failed) instead of the keyspace group discovered by the keyspace name passed from the NewClientWithAPIContextV2 API. 
Although PD (TSO) client almost immediately uses the right keyspace id, queried by keyspace name, to discovery the right keyspace group that the keyspace belongs to, TiDB didn't have retry and failed before PD (TSO) client discovered the right keyspace group service info.
I can consistently repo the issue in staging env with -keyspacename option in pd-tso-bench. 
#### What's the main scenearios hitting the issue?
The standby TiDB pods are provisioned on demand. It seems that in this case the tidb doesn't retry GetTS and surface the problem. We didn't hit this issue in Dev/QA/Unittest, and the reason seems to be the caller of GetTS have retries.  
#### What's the impact
Only impacted the non-default keyspace groups in staging, so it doesn't impact the single-timeline version in eu-central prod. 

### Details
#### What's the problem

**After TiDB calls NewClientWithAPIContextV2 and it returns, the first GetTS always uses the default keyspace group instead of the keyspace group discovered by the keyspace name passed from the NewClientWithAPIContextV2 API. Although PD (TSO) client almost immediately uses the right keyspace id, queried by keyspace name, to discovery the right keyspace group that the keyspace belongs to, TiDB finished the retry within 1ms and failed before PD (TSO) client discovers the right keyspace group service info**

<img width="2297" alt="image" src="https://github.com/tikv/pd/assets/39923490/5fc1c2a8-17e3-47f4-b3b6-ece6b2fc3dba">

```
[2023/07/04 07:09:07.035 +00:00] [INFO] [tso_service_discovery.go:495] ["[tso] updated keyspace group service discovery info"] [keyspaceName=2oVzWDxP4sSxu9f] [keyspaceID=18076] 
    [keyspace-group-service="id:2 user_kind:\"basic\" members:<address:\"http://pd-tso-server-0.tso-service.tidb-serverless.svc:2379\" is_primary:true > members:<address:\"http://pd-tso-server-1.tso-service.tidb-serverless.svc:2379\" > "]
[2023/07/04 07:09:07.035 +00:00] [ERROR] [tso_dispatcher.go:459] ["[tso] getTS error"] [keyspaceName=2oVzWDxP4sSxu9f] [keyspaceID=18076] [dc-location=global] [stream-addr=http://pd-tso-server-1.tso-service.tidb-serverless.svc:2379] [error="[PD:client:ErrClientGetTSO]get TSO failed, after processing requests"]
[2023/07/04 07:09:07.030 +00:00] [INFO] [client.go:522] ["[pd] create pd client with endpoints and keyspace"] [keyspaceName=2oVzWDxP4sSxu9f] [keyspaceID=18076] [pd-address="[serverless-cluster-pd.tidb-serverless.svc:2379]"] [keyspace-name=2oVzWDxP4sSxu9f] [keyspace-id=18076]
[2023/07/04 07:09:07.024 +00:00] [INFO] [tso_service_discovery.go:495] ["[tso] updated keyspace group service discovery info"] [keyspaceName=2oVzWDxP4sSxu9f] [keyspaceID=18076] 
    [keyspace-group-service="user_kind:\"basic\" members:<address:\"http://pd-tso-server-0.tso-service.tidb-serverless.svc:2379\" > members:<address:\"http://pd-tso-server-1.tso-service.tidb-serverless.svc:2379\" is_primary:true > "]
[2023/07/04 07:09:07.019 +00:00] [INFO] [tso_service_discovery.go:584] ["update tso server addresses"] [keyspaceName=2oVzWDxP4sSxu9f] [keyspaceID=18076] [addrs="[http://pd-tso-server-0.tso-service.tidb-serverless.svc:2379,http://pd-tso-server-1.tso-service.tidb-serverless.svc:2379]"]
[2023/07/04 07:09:07.019 +00:00] [INFO] [tso_service_discovery.go:195] ["initializing tso service discovery"] [keyspaceName=2oVzWDxP4sSxu9f] [keyspaceID=18076] [max-retry-times=100] [retry-interval=1s]
```

#### Root Cause Analysis
**newClientWithKeyspaceName**
<img width="1014" alt="image" src="https://github.com/tikv/pd/assets/39923490/cc6efbed-9a5e-4eac-9a23-c51eb8b032fd">
We should synchronously discovery keyspace group service with the updated keyspace ID at line 516 in the above picture.

--- a/client/client.go
+++ b/client/client.go
@@ -277,14 +277,6 @@ type serviceModeKeeper struct {
 	tsoSvcDiscovery ServiceDiscovery
 }
 
-func (k *serviceModeKeeper) SetKeyspaceID(keyspaceID uint32) {
-	k.Lock()
-	defer k.Unlock()
-	if k.serviceMode == pdpb.ServiceMode_API_SVC_MODE {
-		k.tsoSvcDiscovery.SetKeyspaceID(keyspaceID)
-	}
-}
-
 func (k *serviceModeKeeper) close() {
 	k.Lock()
 	defer k.Unlock()
@@ -392,7 +384,7 @@ func createClientWithKeyspace(
 
 	c.pdSvcDiscovery = newPDServiceDiscovery(
 		clientCtx, clientCancel, &c.wg, c.setServiceMode,
-		keyspaceID, c.svrUrls, c.tlsCfg, c.option)
+		nil, keyspaceID, c.svrUrls, c.tlsCfg, c.option)
 	if err := c.setup(); err != nil {
 		c.cancel()
 		return nil, err
@@ -504,23 +496,27 @@ func newClientWithKeyspaceName(
 		opt(c)
 	}
 
+	updateKeyspaceIDCb := func() error {
+		if err := c.initRetry(c.loadKeyspaceMeta, keyspaceName); err != nil {
+			return err
+		}
+		// c.keyspaceID is the source of truth for keyspace id.
+		c.pdSvcDiscovery.(*pdServiceDiscovery).SetKeyspaceID(c.keyspaceID)
+		return nil
+	}
+
 	// Create a PD service discovery with null keyspace id, then query the real id wth the keyspace name,
 	// finally update the keyspace id to the PD service discovery for the following interactions.
 	c.pdSvcDiscovery = newPDServiceDiscovery(
-		clientCtx, clientCancel, &c.wg, c.setServiceMode, nullKeyspaceID, c.svrUrls, c.tlsCfg, c.option)
+		clientCtx, clientCancel, &c.wg, c.setServiceMode, updateKeyspaceIDCb, nullKeyspaceID, c.svrUrls, c.tlsCfg, c.option)
 	if err := c.setup(); err != nil {
 		c.cancel()
 		return nil, err
 	}
-	if err := c.initRetry(c.loadKeyspaceMeta, keyspaceName); err != nil {
-		return nil, err
-	}
-	// We call "c.pdSvcDiscovery.SetKeyspaceID(c.keyspaceID)" after service mode already switching to API mode
-	// and tso service discovery already initialized, so here we need to set the tso_service_discovery's keyspace id too.
-	c.pdSvcDiscovery.SetKeyspaceID(c.keyspaceID)
-	c.serviceModeKeeper.SetKeyspaceID(c.keyspaceID)
 	log.Info("[pd] create pd client with endpoints and keyspace",
-		zap.Strings("pd-address", svrAddrs), zap.String("keyspace-name", keyspaceName), zap.Uint32("keyspace-id", c.keyspaceID))
+		zap.Strings("pd-address", svrAddrs),
+		zap.String("keyspace-name", keyspaceName),
+		zap.Uint32("keyspace-id", c.keyspaceID))
 	return c, nil
 }
 

--- a/client/pd_service_discovery.go
+++ b/client/pd_service_discovery.go
@@ -59,8 +59,6 @@ type ServiceDiscovery interface {
 	GetClusterID() uint64
 	// GetKeyspaceID returns the ID of the keyspace
 	GetKeyspaceID() uint32
-	// SetKeyspaceID sets the ID of the keyspace
-	SetKeyspaceID(keyspaceID uint32)
 	// GetKeyspaceGroupID returns the ID of the keyspace group
 	GetKeyspaceGroupID() uint32
 	// DiscoverServiceURLs discovers the microservice with the specified type and returns the server urls.
@@ -99,6 +97,7 @@ type ServiceDiscovery interface {
 	AddServiceAddrsSwitchedCallback(callbacks ...func())
 }
 
+type updateKeyspaceIDFunc func() error
 type tsoLocalServAddrsUpdatedFunc func(map[string]string) error
 type tsoGlobalServAddrUpdatedFunc func(string) error
 
@@ -149,8 +148,9 @@ type pdServiceDiscovery struct {
 	cancel    context.CancelFunc
 	closeOnce sync.Once
 
-	keyspaceID uint32
-	tlsCfg     *tlsutil.TLSConfig
+	updateKeyspaceIDCb updateKeyspaceIDFunc
+	keyspaceID         uint32
+	tlsCfg             *tlsutil.TLSConfig
 	// Client option.
 	option *option
 }
@@ -160,6 +160,7 @@ func newPDServiceDiscovery(
 	ctx context.Context, cancel context.CancelFunc,
 	wg *sync.WaitGroup,
 	serviceModeUpdateCb func(pdpb.ServiceMode),
+	updateKeyspaceIDCb updateKeyspaceIDFunc,
 	keyspaceID uint32,
 	urls []string, tlsCfg *tlsutil.TLSConfig, option *option,
 ) *pdServiceDiscovery {
@@ -169,6 +170,7 @@ func newPDServiceDiscovery(
 		cancel:              cancel,
 		wg:                  wg,
 		serviceModeUpdateCb: serviceModeUpdateCb,
+		updateKeyspaceIDCb:  updateKeyspaceIDCb,
 		keyspaceID:          keyspaceID,
 		tlsCfg:              tlsCfg,
 		option:              option,
@@ -191,6 +193,14 @@ func (c *pdServiceDiscovery) Init() error {
 		return err
 	}
 	log.Info("[pd] init cluster id", zap.Uint64("cluster-id", c.clusterID))
+
+	// We need to update the keyspace ID before we discover and update the service mode
+	// so that TSO in API mode can be initialized with the correct keyspace ID.
+	if c.updateKeyspaceIDCb != nil {
+		if err := c.updateKeyspaceIDCb(); err != nil {
+			return err
+		}
+	}
 
 	if err := c.checkServiceModeChanged(); err != nil {
 		log.Warn("[pd] failed to check service mode and will check later", zap.Error(err))

--- a/client/tso_service_discovery.go
+++ b/client/tso_service_discovery.go
@@ -277,11 +277,6 @@ func (c *tsoServiceDiscovery) GetKeyspaceID() uint32 {
 	return c.keyspaceID.Load()
 }
 
-// SetKeyspaceID sets the ID of the keyspace
-func (c *tsoServiceDiscovery) SetKeyspaceID(keyspaceID uint32) {
-	c.keyspaceID.Store(keyspaceID)
-}
-
 // GetKeyspaceGroupID returns the ID of the keyspace group. If the keyspace group is unknown,
 // it returns the default keyspace group ID.
 func (c *tsoServiceDiscovery) GetKeyspaceGroupID() uint32 {
@@ -429,12 +424,16 @@ func (c *tsoServiceDiscovery) updateMember() error {
 		return err
 	}
 
+	keyspaceID := c.GetKeyspaceID()
 	var keyspaceGroup *tsopb.KeyspaceGroup
 	if len(tsoServerAddr) > 0 {
-		keyspaceGroup, err = c.findGroupByKeyspaceID(c.GetKeyspaceID(), tsoServerAddr, updateMemberTimeout)
+		keyspaceGroup, err = c.findGroupByKeyspaceID(keyspaceID, tsoServerAddr, updateMemberTimeout)
 		if err != nil {
 			if c.tsoServerDiscovery.countFailure() {
-				log.Error("[tso] failed to find the keyspace group", errs.ZapError(err))
+				log.Error("[tso] failed to find the keyspace group",
+					zap.Uint32("keyspace-id-in-request", keyspaceID),
+					zap.String("tso-server-addr", tsoServerAddr),
+					errs.ZapError(err))
 			}
 			return err
 		}
@@ -448,6 +447,8 @@ func (c *tsoServiceDiscovery) updateMember() error {
 		c.printFallbackLogOnce.Do(func() {
 			log.Warn("[tso] no tso server address found,"+
 				" fallback to the legacy path to discover from etcd directly",
+				zap.Uint32("keyspace-id-in-request", keyspaceID),
+				zap.String("tso-server-addr", tsoServerAddr),
 				zap.String("discovery-key", c.defaultDiscoveryKey))
 		})
 		addrs, err := c.discoverWithLegacyPath()
@@ -487,6 +488,8 @@ func (c *tsoServiceDiscovery) updateMember() error {
 		if primarySwitched := !strings.EqualFold(primaryAddr, c.getPrimaryAddr()); primarySwitched {
 			if _, err := c.GetOrCreateGRPCConn(primaryAddr); err != nil {
 				log.Warn("[tso] failed to connect the next primary",
+					zap.Uint32("keyspace-id-in-request", keyspaceID),
+					zap.String("tso-server-addr", tsoServerAddr),
 					zap.String("next-primary", primaryAddr), errs.ZapError(err))
 				return err
 			}
@@ -497,6 +500,8 @@ func (c *tsoServiceDiscovery) updateMember() error {
 		c.keyspaceGroupSD.update(keyspaceGroup, primaryAddr, secondaryAddrs, addrs)
 	if primarySwitched {
 		log.Info("[tso] updated keyspace group service discovery info",
+			zap.Uint32("keyspace-id-in-request", keyspaceID),
+			zap.String("tso-server-addr", tsoServerAddr),
 			zap.String("keyspace-group-service", keyspaceGroup.String()))
 		if err := c.afterPrimarySwitched(oldPrimary, primaryAddr); err != nil {
 			return err

